server/sonar-server/src/main/java/org/sonar/server/authentication/JwtCsrfVerifier.java

```diff
@@ -0,0 +1,102 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.authentication;
+
+import static org.apache.commons.codec.digest.DigestUtils.sha256Hex;
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+import com.google.common.collect.ImmutableSet;
+import java.math.BigInteger;
+import java.security.SecureRandom;
+import java.util.Set;
+import javax.annotation.Nullable;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.sonar.api.platform.Server;
+import org.sonar.server.exceptions.UnauthorizedException;
+
+public class JwtCsrfVerifier {
+
+  private static final String CSRF_STATE_COOKIE = "XSRF-TOKEN";
+  private static final String CSRF_HEADER = "X-XSRF-TOKEN";
+
+  private static final Set<String> UPDATE_METHODS = ImmutableSet.of("POST", "PUT", "DELETE");
+  private static final String API_URL = "/api";
+  private static final Set<String> RAILS_UPDATE_API_URLS = ImmutableSet.of(
+    "/api/events",
+    "/api/favourites",
+    "/api/issues/add_comment",
+    "/api/issues/delete_comment",
+    "/api/issues/edit_comment",
+    "/api/issues/bulk_change",
+    "/api/projects/create",
+    "/api/properties/create",
+    "/api/server",
+    "/api/user_properties"
+  );
+
+  private final Server server;
+
+  public JwtCsrfVerifier(Server server) {
+    this.server = server;
+  }
+
+  public String generateState(HttpServletResponse response, int timeoutInSeconds) {
+    // Create a state token to prevent request forgery.
+    // Store it in the cookie for later validation.
+    String state = new BigInteger(130, new SecureRandom()).toString(32);
+
+    Cookie cookie = new Cookie(CSRF_STATE_COOKIE, sha256Hex(state));
+    cookie.setPath(server.getContextPath() + "/");
+    cookie.setSecure(server.isSecured());
+    cookie.setHttpOnly(false);
+    cookie.setMaxAge(timeoutInSeconds);
+    response.addCookie(cookie);
+    return state;
+  }
+
+  public void verifyState(HttpServletRequest request, @Nullable String csrfState) {
+    if (!shouldRequestBeChecked(request)) {
+      return;
+    }
+    String stateInHeader = request.getHeader(CSRF_HEADER);
+    if (isBlank(csrfState) || isBlank(stateInHeader) || !sha256Hex(csrfState).equals(stateInHeader)) {
+      throw new UnauthorizedException();
+    }
+  }
+
+  private static boolean shouldRequestBeChecked(HttpServletRequest request) {
+    String path = request.getRequestURI().replaceFirst(request.getContextPath(), "");
+    return UPDATE_METHODS.contains(request.getMethod())
```

Comment:
optimization: as most requests are GET, checking the method should be done before extracting the variable `path` from `request.getRequestURI()`